stingray/pulse/pulsar.py

```diff
@@ -303,7 +304,107 @@ def profile_stat(profile, err=None):
     return np.sum((profile - mean) ** 2 / err ** 2)
 
 
-def z_n(phase, n=2, norm=1):
+@functools.lru_cache(maxsize=128)
+def _cached_sin_harmonics(nbin, z_n_n):
+    dph = 1.0 / nbin
+    twopiphases = np.pi * 2 * np.arange(dph / 2, 1, dph)
+    cached_sin = np.zeros(z_n_n * nbin)
+    for i in range(z_n_n):
+        cached_sin[i * nbin: (i + 1) * nbin] = np.sin(twopiphases)
+    return cached_sin
+
+
+@functools.lru_cache(maxsize=128)
+def _cached_cos_harmonics(nbin, z_n_n):
+    dph = 1.0 / nbin
+    twopiphases = np.pi * 2 * np.arange(dph / 2, 1, dph)
+    cached_cos = np.zeros(z_n_n * nbin)
+    for i in range(z_n_n):
+        cached_cos[i * nbin: (i + 1) * nbin] = np.cos(twopiphases)
+    return cached_cos
+
+
+@jit(nopython=True)
+def _z_n_fast_cached_sums_unnorm(prof, ks, cached_sin, cached_cos):
+    all_zs = np.zeros(ks.size)
+    N = prof.size
+
+    total_sum = 0
+    for k in ks:
+        local_z = (
+            np.sum(cached_cos[: N * k: k] * prof) ** 2
+            + np.sum(cached_sin[: N * k: k] * prof) ** 2
+        )
+        total_sum += local_z
+        all_zs[k - 1] = total_sum
+
+    return all_zs
+
+
+def z_n_poisson_all(profile, nmax=20):
+    '''Z^2_n statistic for multiple harmonics and Poissonian pulse profiles
+
+    See Bachetti+2021, arXiv:2012.11397
+
+    Parameters
+    ----------
+    profile : array of floats
+        The folded pulse profile (containing the number of
+        photons falling in each pulse bin)
+    n : int
+        Number of harmonics, including the fundamental
+
+    Returns
+    -------
+    ks : list of ints
+        Harmonic numbers, from 1 to nmax (included)
+    z2_n : float
+        The value of the statistic for all ks
+    '''
+    cached_sin = _cached_sin_harmonics(profile.size, nmax)
+    cached_cos = _cached_cos_harmonics(profile.size, nmax)
+    ks = np.arange(1, nmax + 1, dtype=int)
+
+    total = np.sum(profile)
+    if total == 0:
+        return ks, np.zeros(nmax)
+    all_zs = _z_n_fast_cached_sums_unnorm(profile, ks, cached_sin, cached_cos)
+
+    return ks, all_zs * 2 / total
+
+
+def z_n_gauss_all(profile, err, nmax=20):
+    '''Z^2_n statistic for multiple harmonics and Gaussian pulse profiles
```

Comment:
small nitpick: pulse profile with Gaussian flux uncertainties. Otherwise a user might think that the pulse profile shape itself is Gaussian.

Reply:
Done